src/common/base32.h

```diff
@@ -0,0 +1,82 @@
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+/**
+ *  Copyright (C) 2015 Trustifier Inc.
+ *  Copyright (C) 2015 Ahmed Masud
+ *  Copyright (C) 2015 Topology LP
+ *  Copyright (C) 2018 Jakob Petsovits
+ *  All rights reserved.
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining a copy
+ *  of this software and associated documentation files (the "Software"), to
+ *  deal in the Software without restriction, including without limitation the
+ *  rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ *  sell copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in
+ *  all copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ *  THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ *  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ *  IN THE SOFTWARE.
+ *
+ *  Adapted from https://github.com/ahmed-masud/libbase32,
+ *  commit 79761b2b79b0545697945efe0987a8d3004512f9.
+ *  Quite different now.
+ */
+
+#include <string>
+
+
+namespace tools
+{
+namespace base32
+{
+
+using alphabet_index_t = uint_fast16_t;
+
+void init(std::string& result, size_t capacity);
+size_t decoded_max_size(size_t encoded_size) noexcept;
+size_t encoded_size(size_t binary_size) noexcept;
+void decode_block(std::string& decoded, const alphabet_index_t* idx);
+void decode_tail(std::string& decoded, const alphabet_index_t* idx, size_t idx_len);
```

Comment:
Do you happen to know what "decode the tail" means, and whether we really need this?

Reply:
Yes we need it. It decodes correctly the last remaining block. You can check the idea of the algorithm here:
https://herongyang.com/Encoding/Base32-Encoding-Algorithm.html

I believe that this is the typical PR that needs different implementations, comparison and performance tests. I dont expect this exact code to be used in the final version but so far I do believe that it is correct, optimized and will serve for our purposes at the moment.